text/3323-restrictions.md

```diff
@@ -0,0 +1,433 @@
+Trait :
+   unsafe?
++  ImplRestriction?
+   trait IDENTIFIER
+   GenericParams? ( : TypeParamBounds? )? WhereClause? {
+     InnerAttribute*
+     AssociatedItem*
+   }
+
++ImplRestriction :
++   impl ( crate )
++   | impl ( self )
++   | impl ( super )
++   | impl ( in SimplePath )
+```
+
+Essentially, `mut` and `impl` have the same syntax as `pub`, just with a different keyword. Using
+the keyword without providing a path is not allowed.
+
+## Behavior
+
+The current behavior of `pub` is that `pub` makes something visible within the declared scope. If no
+scope is declared (such that it is just `pub`), then the item is visible everywhere. This behavior
+is preserved for `impl` and `mut`. When a restriction is used, the behavior is allowed only within
+the declared scope. While in most cases the default visibility is private, `pub` is default in some
+cases, namely `enum` variants, `enum` fields, and `trait` items. `impl` and `mut` will have a
+consistent default: when omitted entirely, the scope is inherited from `pub`. This is both what is
+most convenient and is what is required for backwards compatibility with existing code.
+
+When an `ImplRestriction` is present, implementations of the associated trait are only permitted
+within the designated path. Any implementation of the trait outside this scope is a compile error.
+When a `MutRestriction` is present, mutable uses of the associated field are only permitted within
+the designated path. Any mutable use of the field outside the scope is a compile error. Further, a
+`struct`, `union`, or `enum` variant containing fields with an associated `MutRestriction` may not
+be constructed with `struct` expressions unless all fields are unrestricted in the present scope.
+This is the case even if the field is not directly declared, such as when functional record updates
+are used.
+
+## "Mutable use" in the compiler
+
+The concept of a "mutable use" [already exists][mutating use method] within the compiler. This
+catches all situations that are relevant here, including `ptr::addr_of_mut!`, `&mut`, and direct
+assignment to a field, while excluding interior mutability. As such, formal semantics of what
+constitutes a "mutable use" are not stated here.
+
+[mutating use method]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/visit/enum.PlaceContext.html#method.is_mutating_use
+
+## Interaction with `trait` aliases
+
+Trait aliases cannot be implemented. As such, there is no concern about compatibility between the
+`impl` restriction and `trait` aliases.
+
+# Drawbacks
+
+- Additional syntax for macros to handle
+
+# Alternatives
+
+- `impl` and `mut` restrictions could be attributes, similar to `#[non_exhaustive]`.
+  - The proposed syntax could by syntactic sugar for these attributes.
+- Visibility could be altered to accept restrictions as a type of parameter, such as
+  `pub(crate, mut = self)`. This does not work because restrictions are not permitted everywhere
+  visibility is.
```

Comment:
I'm not a huge fan of this alternative, but it *could* work, just make it an error to use a restriction in a visability modifier in some places (yeah, it's not awesome, but it would work, I think).

Reply:
That just needlessly pushes the error farther into the compilation process, no?

Reply:
Sure, I don't deny there are downsides, but this is just input into a trade-off decision, it doesn't mean that this approach "does not work".

Reply:
I'll update the text.